euphoria-spark/src/main/java/cz/seznam/euphoria/spark/BatchJoinIterator.java

```diff
@@ -143,4 +165,57 @@ private void emitLeft() {
   private boolean sameKey(Tuple2<BatchJoinKey<K>, ?> a, Tuple2<BatchJoinKey<K>, ?> b) {
     return Objects.equals(a._1.getKey(), b._1.getKey());
   }
+
+  private void addElementToStats(Tuple2<BatchJoinKey<K>, Either<L, R>> element){
+
+    K key = element._1.getKey();
+
+    Tuple2<MutableLong, MutableLong> leftRightCounts = numOfElementsByKey
+        .computeIfAbsent(key, (k) -> new Tuple2<>(new MutableLong(), new MutableLong()));
+
+    Either<L, R> eitherSide = element._2;
+
+    if(eitherSide.isLeft()){
+      leftRightCounts._1.increment();
+    }else {
+      leftRightCounts._2.increment();
+    }
+
+    numOfEncounteredElements++;
+  }
+
+  private void logStats(){
+
+    if(!LOG.isInfoEnabled()){
+      return;
+    }
+
+    LOG.info("-- {} statistics:", BatchJoinIterator.class.getSimpleName());
+    LOG.info("-- Keys count: {}, input elements count: {}, output (emitted/joined) elements #: {}.",
+        numOfElementsByKey.size(), numOfEncounteredElements, numOfEmittedElements);
+
+    List<Entry<K, Tuple2<MutableLong, MutableLong>>> top10Keys = numOfElementsByKey.entrySet()
```

Comment:
this can introduce a memory problem, it basically means that we need to fit all keys for a single task in memory. Can we keep track of top N elements using priority queue?

Reply:
That would mean that we need to fit only N keys in memory

Reply:
I believe that we can keep top N elements only when pre-grupped by key. Or in another words keys coming from wrapped iterators do not change until all elements with the same key from input dataset are depleted. Does this assumption hold? It should when we are doing join, but I was not entirely sure.
I will re implement it soon.

Reply:
yes, all elements always come sorted by KEY and secondary sorted by SIDE (left side first) ;)